src/node_env_var.cc

```diff
@@ -66,89 +66,76 @@ std::shared_ptr<KVStore> system_environment = std::make_shared<RealEnvStore>();
 Local<String> RealEnvStore::Get(Isolate* isolate,
                                 Local<String> property) const {
   Mutex::ScopedLock lock(per_process::env_var_mutex);
-#ifdef __POSIX__
+
   node::Utf8Value key(isolate, property);
-  const char* val = getenv(*key);
-  if (val) {
-    return String::NewFromUtf8(isolate, val, NewStringType::kNormal)
-        .ToLocalChecked();
-  }
-#else  // _WIN32
-  node::TwoByteValue key(isolate, property);
-  WCHAR buffer[32767];  // The maximum size allowed for environment variables.
-  SetLastError(ERROR_SUCCESS);
-  DWORD result = GetEnvironmentVariableW(
-      reinterpret_cast<WCHAR*>(*key), buffer, arraysize(buffer));
-  // If result >= sizeof buffer the buffer was too small. That should never
-  // happen. If result == 0 and result != ERROR_SUCCESS the variable was not
-  // found.
-  if ((result > 0 || GetLastError() == ERROR_SUCCESS) &&
-      result < arraysize(buffer)) {
-    const uint16_t* two_byte_buffer = reinterpret_cast<const uint16_t*>(buffer);
-    v8::MaybeLocal<String> rc = String::NewFromTwoByte(
-        isolate, two_byte_buffer, NewStringType::kNormal);
-    if (rc.IsEmpty()) {
-      isolate->ThrowException(ERR_STRING_TOO_LONG(isolate));
+  char* val = nullptr;
+  size_t init_sz = 256;
+
+  // Allocate 256 bytes initially, if not enough reallocate.
+  val = Malloc(init_sz);
+
+  int ret = uv_os_getenv(*key, val, &init_sz);
+
+  if (UV_ENOBUFS == ret) {
+    // Buffer is not large enough, reallocate to the updated init_sz
+    // and fetch env value again.
+    val = Realloc(val, init_sz);
+
+    ret = uv_os_getenv(*key, val, &init_sz);
+
+    // Still failed to fetch env value return emptry string.
+    if (UV_ENOBUFS == ret || UV_ENOENT == ret) {
       return Local<String>();
```

Comment:
This could be a memory leak, because we do not free `val`.

We typically use RAII-style constructs for managing memory that automatically deallocate their resources when the function in which they are is left; for example, `std::vector` would work, or `MaybeStackBuffer` (which would fit in nicely with the 256-byte-default that you have – with `MaybeStackBuffer`, we could avoid allocating memory altogether in the most common cases).

Reply:
fixed @ https://github.com/nodejs/node/pull/27310/commits/d6db97ed609a1b3d33a2ae55c0e3e0ed81b221bb